add entry_points field on python_distribution

kaos opened the pull request:
-  add new `entry_points` field for the `python_distribution()` target.
-  support dependency inference on `entry_points`.
-  support `pex_binary` targets in entry points, which will use the `entry_point` from the `pex_binary()` target.

The `entry_points` field of `python_distribution` is a `dict[str, dict[str, str]]` value (just as for `setuptools.setup()`), where the `str` value is either on the form of a regular setuptools entry_point or a pants target address. Any value that either starts with a `:` or that contains a `/` is considered a target address (for a `pex_binary`) and resolves to the binary's entry_point.

The reason for not requiring a leading `//` for target addresses on the entry_point str value, is to support the same syntax 1:1 from the entry_point value on `.with_binaries()`, to ease migration.

Example `BUILD` (from this repo)
```python
python_distribution(
  name='pants-packaged',
  ...
  provides=setup_py(
    name='pantsbuild.pants',
    ...
  ),
  entry_points={
    'console_scripts': {
      'pants': 'src/python/pants/bin:pants',
    },
  },
)
```

Example `BUILD`, before this change:
```python
... 
  provides=setup_py(
    name='pantsbuild.pants',
    ...
  ).with_binaries(
    pants='src/python/pants/bin:pants',
  )
```

fixes #11807.
fixes #11808.
closes #11880

Signed-off-by: Andreas Stenius <andreas.stenius@svenskaspel.se>

[ci skip-rust]

[ci skip-build-wheels]

kaos commented:
I've left three `# type: ignore` in there, that I'm not really sure how to handle. That caused these errors:

```
"Optional[FrozenDict[str, FrozenDict[str, EntryPoint]]]", base class
"NestedDictStringToStringField" defined the type as
"Optional[FrozenDict[str, FrozenDict[str, str]]]")  [assignment]
        value: Optional[FrozenDict[str, FrozenDict[str, EntryPoint]]]
        ^
src/python/pants/backend/python/target_types.py:703:81: error: Incompatible
types in assignment (expression has type
"Optional[FrozenDict[str, FrozenDict[str, EntryPoint]]]", base class
"NestedDictStringToStringField" defined the type as
"Optional[FrozenDict[str, FrozenDict[str, str]]]")  [assignment]
    ...lassVar[Optional[FrozenDict[str, FrozenDict[str, EntryPoint]]]] = None
                                                                         ^
src/python/pants/backend/python/target_types.py:713:5: error: Return type
"Optional[FrozenDict[str, FrozenDict[str, EntryPoint]]]" of "compute_value"
incompatible with return type "Optional[FrozenDict[str, FrozenDict[str, str]]]"
in supertype "NestedDictStringToStringField"  [override]
        def compute_value(
        ^
Found 3 errors in 1 file (checked 455 source files)
```

Also, I've not done any dependency inference here yet.. either this will be good enough without, and we can address that in #11807 or we resolve both issues in this PR.

kaos commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -545,12 +546,20 @@ async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
         Get(ResolvedPexEntryPoint, ResolvePexEntryPointRequest, request)
         for request in entry_point_requests
     )
+    entry_points = setup_kwargs.setdefault("entry_points", {})
```
This will unconditionally add `entry_points` to the generated setup.py file. Maybe not such a big deal.. ?
I did leave the console_scripts in the for-loop on purpose, as we don't need that outside the for-loop.

jsirois replied:
It will probably be friendly to future debugging sessions to leave it out if there are none.

How about just `entry_points = defaultdict(list)` to simplify the for loops a bit and then one line of post-processing that says:
```python
if entry_points:
  setup_kwargs["entry_points"] = entry_points
```

```

kaos commented on src/python/pants/backend/python/goals/setup_py_test.py:
```diff
@@ -227,6 +227,129 @@ def test_generate_chroot(chroot_rule_runner: RuleRunner) -> None:
     )
 
 
+def test_generate_chroot_entry_points(chroot_rule_runner: RuleRunner) -> None:
```
Copied the other chroot test so this one could be trimmed down to only test the entry points stuff. I like tests that doesn't test too much, but if you feel like keeping it all in a single test, I'm ok with that too.. ;)

jsirois replied:
Yeah, if you're willing, trimming down to just setup entrypoints would be great. Certainly the .js and .pyi treatment is not needed in this test.

kaos commented on src/python/pants/backend/python/macros/python_artifact.py:
```diff
@@ -1,7 +1,31 @@
+from pants.base.deprecated import deprecated
+
+
+def _check_entry_points(
```
To make it easier to work with the entry points else where, it is beneficial to have a single format to expect. This one translates the user input to a predictable format.

Eric-Arellano replied:
Sg! Name makes this sound like it's only validation though. How about `_normalize_entry_points`?

jsirois commented on src/python/pants/BUILD:
```diff
@@ -34,9 +34,12 @@ python_distribution(
+  ),
+  entry_points={
+    'console_scripts': {
+      'pants': 'pants.bin.pants_loader:main',
```
I do think it would be great to circle back and add rule support for parsing this field to figure out if it is a target address to support the old DRY `.with_binaries` provided. Relative address syntax (`:name`) is unambiguous, but otherwise we could just require `//explicit/absolute:form` and the `//` would make everything unambiguous. Perhaps add a TODO / tracking issue link if you agree.

Eric-Arellano replied:
If we require the prefix of `//`, then I'm okay with it. As explained in https://github.com/pantsbuild/pants/issues/11808#issuecomment-811358327, I bias towards the simplicity of only having the module format. But requiring the unambiguous syntax does resolve half of my concern. 

The only remaining concern now is that we have two ways of doing something, which means more documentation etc. But that isn't too concerning:

> Blah blah...should be in the format `path.to.module` or `path.to.module:func`. You can also point to a `pex_binary` target to use its `entry_point` field by using either `:sibling_target` or `//path/to:target`.

What do you think @kaos?

kaos replied:
I'd like to keep it. It helps with DRYness, and also makes migration from `.with_binaries()` easier. 
The only downside I see is if it adds another difference in semantics to `entry_points` on `python_distribution` vs `setup_py()`.

Maybe deprecate `entry_points` on `setup_py()` and flag it as a reserved field (akin to packages and a few others), to get rid of the potential pitfalls.

Eric-Arellano replied:
Sg on allowing pex_binary addresses if it's unambiguous by either `:tgt` or `//path/to:tgt` (`//path/to` is also valid, where the target's name == parent folder). If it does not start with either `//` or `:`, we assume it's a module.

> Maybe deprecate entry_points on setup_py() and flag it as a reserved field (akin to packages and a few others), to get rid of the potential pitfalls.

I don't think we need to do that - you can still use the old `setup_py()` way of doing it, only you'll miss out a) dependency inference, and b) being able to use a `pex_binary` address. By allowing `setup_py()`, we reduce possible churn and it allows people to dynamically compute the value with the plugin hook to generate `setup_py()` args: https://www.pantsbuild.org/docs/plugins-setup-py

kaos replied:
I opted for a check that if it starts with `:` or _contains_ any `/` we treat it as a target address, as this was the supported syntax for `.with_binaries()`, to ease migrations, but am open to switch to the stricter check if you prefer that.

jsirois commented on src/python/pants/backend/python/goals/setup_py_test.py:
```diff
@@ -227,6 +227,129 @@ def test_generate_chroot(chroot_rule_runner: RuleRunner) -> None:
+            """
+            python_distribution(
+                name='foo-dist',
+                # TODO: remove this once dep inference from entry points works
```
Is there a tracking issue you can reference here?

kaos replied:
Yes, #11807

kaos replied:
Resolved this in this PR.

jsirois commented on src/python/pants/backend/python/goals/setup_py_test.py:
```diff
@@ -227,6 +227,129 @@ def test_generate_chroot(chroot_rule_runner: RuleRunner) -> None:
+                            "foo_qux":"foo.baz.qux",
+                        },
+                        "foo_plugins":[
+                            "foo-bar=foo.bar:plugin",
```
Testing and handling collision would be great. Presumably the right thing to do is to fail to alert the BUILD file author to simplify.

kaos replied:
this case is covered in test_merge_entry_points.

jsirois reviewed (commented):
Thanks @kaos. Sorry for the delay here. I may have missed some context since Eric was working more closely with you on this, so please push back if anything I've pointed out has already been worked out in contradiction.

Eric-Arellano commented on src/python/pants/backend/python/macros/python_artifact.py:
```diff
@@ -1,7 +1,31 @@
+def _check_entry_points(
+    entry_points: Optional[Dict[str, Union[List[str], Dict[str, str]]]]
+) -> Optional[Dict[str, List[str]]]:
+    """Ensure any entry points are on the form a dictionary of string -> list of strings."""
```
Grammar nit:

```suggestion
    """Ensure any entry points are in the form a dictionary of string -> list of strings."""
```

Eric-Arellano commented on src/python/pants/backend/python/macros/python_artifact.py:
```diff
@@ -21,6 +45,12 @@ def __init__(self, **kwargs):
                 f"{type(name)}."
             )
 
+        # coerce entry points from dict of string -> dict of string -> string to
+        # dict of string -> list of string
```
May be easier to read this as `dict[str, dict[str, str]]` -> `dict[str, list[str]]`.

Eric-Arellano commented on src/python/pants/backend/python/macros/python_artifact.py:
```diff
@@ -40,6 +70,38 @@ def binaries(self):
+      }
+    }
+
+The entry point must now be provided explicitly and are not derived from a
+`pex_binary` target.
```
If you think that we should support pex_binary addresses via John's unambiguous syntax, it would be great to include in this PR so that this deprecation can be fixed.

kaos replied:
dropped deprecation for later.

Eric-Arellano commented on src/python/pants/backend/python/macros/python_artifact.py:
```diff
@@ -40,6 +70,38 @@ def binaries(self):
 
+    @deprecated(
+        "2.6.0dev0",
+        """Use `python_distribution(entry_points={'console_scripts':{'<name>': '<entry
```
Masterful deprecation message 🥰

Eric-Arellano commented on src/python/pants/engine/target.py:
```diff
@@ -1179,6 +1179,35 @@ def compute_value(
         return FrozenDict(value_or_default)
 
 
+class NestedDictStringToStringField(Field):
```
Sounds reasonable for this to be in `target.py`. Thanks!

Eric-Arellano commented on src/python/pants/engine/target.py:
```diff
@@ -1179,6 +1179,35 @@ def compute_value(
+            address,
+            cls.alias,
+            raw_value,
+            expected_type="a dictionary of string -> a dictionary of string -> string",
```
Ditto that it may be easier to grok this as a type hint: `dict[str, dict[str, str]]`.

kaos replied:
done

Eric-Arellano commented on src/python/pants/engine/target_test.py:
```diff
@@ -669,6 +670,45 @@ class ExampleDefault(DictStringToStringField):
     assert ExampleDefault(None, address=addr).value == FrozenDict({"default": "val"})
 
 
+def test_nested_dict_string_to_string_field() -> None:
```
Thank you for the tests!!

Eric-Arellano reviewed (commented):
Really great stuff! Thank you! If you're willing, I think it would be best to include the dependency inference PR directly in this one. That way, the new deprecated feature is not degrading the experience from before.

You could also maybe leave off the deprecation part of this PR and save for a followup. That way, in the changelog, we would have something like "Add `entry_points` field to `python_distribution`" target (new feature), and then the followup could be "Deprecate `.with_binaries()` in favor of `entry_points` field".

Eric-Arellano commented:
(Also pardon the delay in reviewing this!)

kaos commented:
> (Also pardon the delay in reviewing this!)

Don't worry about it. We all have priorities to consider :)
I'll look at addressing your and Johns feedback in the coming days. Could be until the weekend before I have time to push anything.

kaos commented:
OK, this has been long overdue. Sorry for that (I practically had to start over, as I'd forgot all context in this.. so can only guess you have too)
I've cleaned up according to previous review comments. Dropped the deprecation of the `.with_binaries()` for another PR, and added support for addressing `pex_binary` targets from `python_distribution(entry_points)`.

I opted to go with a little more liberal approach to determine if it is a build target address. As long as there is a slash in the address, it will never be a setuptools entry point (at the risk of them introducing that.. but I think that risk is rather low)
One can always prefer the absolute address path to be sure (and we can add warnings about it in the future, if need be)
The reason was to stay as close as possible to the supported syntax used for target addresses in the `.with_binaries()` construct, as can be seen in the diff of the `src/python/pants/BUILD` file..

It is easy enough to tighten this requirement up, of course (it is just a single if statement doing this check)
https://github.com/pantsbuild/pants/pull/11872/files#diff-4740170d8c26730e2b985731f39b65d20446a0e3ea2def3c82491c3f7c99a17fR158

Dependency inference should work for distribution entry points as well.

Let me know what you think, and I'll fix them up sooner than in another quarter ;)

Eric-Arellano commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -576,15 +580,45 @@ async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
+                    name: entry_point.spec for name, entry_point in section.items()
+                }
+
+    # Collect any entry points from the setup_py() target
```
```suggestion
    # Collect any entry points from the setup_py() object.
```

Eric-Arellano commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -576,15 +580,45 @@ async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
             )
         entry_point_requests.append(ResolvePexEntryPointRequest(binary[PexEntryPointField]))
+
+    entry_point_sources = cast(
+        Dict[str, Dict[str, Dict[str, str]]], recursive_defaultdict_factory()
+    )
```
Rather than defining the dictionary here and appending to it through the next few lines, it might be more clear to have 3 distinct variables `with_binaries_entry_points`, `setup_py_kwargs_entry_points`, and `entry_points_field_entry_points`. That gives better encapsulation and makes clear what each distinct block is doing.

Then, when you call  `_merge_entry_point`, you can use these same descriptions of the source. It'll all be centralized to one call site, and more clear why you're giving a description.

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -135,29 +142,124 @@ async def inject_pex_binary_entry_point_dependency(
+        for name, ref in values.items():
+            if ref.startswith(":") or "/" in ref:
+                targets = await Get(Targets, UnparsedAddressInputs([ref], owning_address=address))
+                if not targets:
+                    raise InvalidFieldException(
+                        f'The "entry_points" target address {repr(ref)} does not resolve to any known target.'
+                    )
```
No need for this. The rule will already error if the address is not resolvable.

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -135,29 +142,124 @@ async def inject_pex_binary_entry_point_dependency(
+                        f'The "entry_points" target address {repr(ref)} does not resolve to any known target.'
+                    )
+
+                for target in targets:
```
There are no globs for addresses, so you can assume `targets` is exactly 1. Fwit, you can add an `AssertionError` if that is not the case.

This will allow you to remove the `for` loop nesting.

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -135,29 +142,124 @@ async def inject_pex_binary_entry_point_dependency(
+            else:
+                ep = EntryPoint.parse(ref, f"{name} for {address} {section}")
+
+            if ep is not None:
```
I don't think `ep` can ever be `None`, right?

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -135,29 +142,124 @@ async def inject_pex_binary_entry_point_dependency(
+    entry_points: dict[str, dict[str, EntryPoint]] = defaultdict(dict)
+
+    for section, values in field_value.items():
+        for name, ref in values.items():
+            if ref.startswith(":") or "/" in ref:
+                targets = await Get(Targets, UnparsedAddressInputs([ref], owning_address=address))
```
We'll want to parallelize this via `MultiGet`. It's an antipattern to use `Get` inside a loop like this: https://www.pantsbuild.org/docs/rules-api-tips#tip-use-multiget-for-increased-concurrency

I'm not certain the best way to do that yet though, and it might be more clear after making below changes. One possible way is to factor out a helper rule which processes a single "section", then use `await MultiGet` where you have a `Get` for each section.

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -135,29 +142,124 @@ async def inject_pex_binary_entry_point_dependency(
+                    )
+
+                for target in targets:
+                    if target.has_field(PexEntryPointField):
+                        binary_entry_point = await Get(
+                            ResolvedPexEntryPoint,
+                            ResolvePexEntryPointRequest(target[PexEntryPointField]),
+                        )
+                        ep = binary_entry_point.val
+                        break
+                else:
+                    raise InvalidFieldException(
+                        f'The "entry_points" target address {repr(ref)} does not resolve to a pex_binary() target, but:'
+                        + ", ".join(str(t) for t in targets)
+                    )
```
As discussed over DM, this has been confusing me a lot on whether you need to use a `pex_binary` or not. 

I think it makes sense to always require `pex_binary`. We can't reliably convert any other target type like `python_library` into a module. We can only do that with `pex_binary` because it has an `entry_point` field.

--

There's additional nuance: the `console_scripts` category `console_scripts` requires the `:func` part to be defined. So, for `.with_binaries()`, we've been requiring that you specify the `:func` part in the original `pex_binary`. I think we want to eagerly validate that here too when the section is `console_scripts`:

https://github.com/pantsbuild/pants/blob/ddc3c3d791178a610a9b8988cb74ac072bdac9c0/src/python/pants/backend/python/goals/setup_py.py#L584-L590

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -135,29 +142,124 @@ async def inject_pex_binary_entry_point_dependency(
+
+    entry_point_addresses = Addresses()
+
+    if entry_points.val and python_infer_subsystem.entry_points:
```
We should have been respecting `python_infer_subsystem.entry_points` from the start, including with `.with_binaries`. You can move this check to the very top of the rule and early return if the option is disabled.

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -135,29 +142,124 @@ async def inject_pex_binary_entry_point_dependency(
+
+    if entry_points.val and python_infer_subsystem.entry_points:
+        address = original_tgt.target.address
+        explicitly_provided_deps = await Get(
+            ExplicitlyProvidedDependencies, DependenciesRequest(original_tgt.target[Dependencies])
+        )
```
It'd be good to do this sooner via a heterogeneous `MultiGet` on the above `Get`: https://www.pantsbuild.org/docs/rules-api-tips#tip-use-multiget-for-increased-concurrency.

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -135,29 +142,124 @@ async def inject_pex_binary_entry_point_dependency(
+
+    with_binaries = original_tgt.target[PythonProvidesField].value.binaries
+    if not with_binaries:
+        pex_addresses = Addresses()
```
`with_binaries_addresses` is more descriptive

Eric-Arellano commented on src/python/pants/backend/python/target_types_test.py:
```diff
@@ -339,35 +341,77 @@ def test_parse_requirements_file() -> None:
+                    name="my-dist-b"
+                ),
+                entry_points={
+                    "console_scripts":{
+                        "b_cmd": "project.app:main"
+                    }
```
You'll also want to check using a `pex_binary` address, can even be the same one used above for `with_binaries`.

Eric-Arellano replied:
Also a dep on `ansicolors`

kaos replied:
not sure what you me to test wrgt `ansicolors`.. ? (how/where?)
I see there is a requirement library for ansicolors in the test, that doesn't seem to do anything.. but don't know how to pull that in, in a useful way.

Eric-Arellano reviewed (commented):
Thanks @kaos! Pardon the delay, I think I now understand what we're after, specifically with `pex_binary` and `console_scripts`. Thanks for the great PR description :)

kaos commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -970,6 +1004,43 @@ def has_args(call_node: ast.Call, required_arg_ids: Tuple[str, ...]) -> bool:
+    """Merge all entry points, throwing ValueError if there are any conflicts."""
+    # keep source_name info for each entry_point until we have merged all
+    # sources, so we can provide better error messages in case of conflicts
+    merged = cast(
+        Dict[str, Dict[str, List[Tuple[str, str]]]], defaultdict(partial(defaultdict, list))
+    )
```
Would you prefer I change it/refactor it, or just explain what in a comment.. ?

kaos replied:
opt-ed for a comment for now.. ;)

kaos commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -970,6 +1004,43 @@ def has_args(call_node: ast.Call, required_arg_ids: Tuple[str, ...]) -> bool:
 
 
+def merge_entry_points(
+    *entry_points: Tuple[str, Dict[str, Dict[str, str]]]
```
I renamed the arg, but kept the type hint, as it comes in as `(desc, entry_points), (desc, entry_points), ...`
it would complicate the code calling this method if we would want to swap the members of the tuples...

kaos commented on src/python/pants/backend/python/target_types_test.py:
```diff
@@ -364,35 +366,90 @@ def test_parse_requirements_file() -> None:
+    assert_injected(
+        Address("project", target_name="dist-b"),
+        Address("project", relative_file_path="app.py", target_name="my_library"),
+        Address("who_knows", relative_file_path="module.py", target_name="random_lib"),
```
Notice a small difference here, compared to `.with_binaries()` here you get a dependency on the owner of the source file that implements the entry point, whereas for `.with_binaries()` you get a dependency on the `pex_binary` itself.
Not sure if this is cause for any issues, down the road.. ?

Eric-Arellano replied:
Oh hm....good eye! You're right that this is a problem. When you have an address to a `pex_binary`, we want a dep on the `pex_binary`, not the `python_library` that has the source code for its entry point.

This will require some changes to your dep inference rule. Right now, you are normalizing the whole `entry_points` field to a set of modules, then doing inference from there. That works when the user specified a module explicitly, but that does not work when they had a target address. Instead, you'll need to differentiate between those two formats when doing inference: if it's a target address, use the same `Get(Addresses, UnparsedAddressInputs)` used for `.with_binaries`, else use the `Get(PythonModuleOwners, PythonModule)` stuff.

> you'll need to differentiate between those two formats when doing inference

Maybe update `ResolvedPythonDistributionEntryPoints` to store two fields: `pex_binary_addresses` and `explicit_modules`, then have a property that merges the two into `all_entry_points`? That allows `setup_py.py` to use `all_entry_points` because it doesn't care about the nuance, but dep inference to use the two fields.

Eric-Arellano commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -589,15 +586,47 @@ async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
+        )
+
+        if entry_points_field.val:
+            for key, section in entry_points_field.val.items():
+                source_entry_points[key] = {
+                    name: entry_point.spec for name, entry_point in section.items()
+                }
```
Could make this a nested dict comprehension, where you redefine `source_entry_points` to the comprehension.

Eric-Arellano commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -589,15 +586,47 @@ async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
         Get(ResolvedPexEntryPoint, ResolvePexEntryPointRequest, request)
         for request in entry_point_requests
     )
+    source_with_binaries = dict()
```
Also, better to use a dict literal `{}` https://www.pantsbuild.org/docs/style-guide#use-collection-literals.

Eric-Arellano commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -589,15 +586,47 @@ async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
+                }
+
+    # Collect any entry points from the setup_py() object.
+    source_provides = dict()
+    if "entry_points" in setup_kwargs:
+        source_provides = setup_kwargs["entry_points"]
```
```suggestion
    entry_points_from_provides = setup_kwargs.get("entry_points", {})
```

Eric-Arellano commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -589,15 +586,47 @@ async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
+                    name: entry_point.spec for name, entry_point in section.items()
+                }
+
+    # Collect any entry points from the setup_py() object.
```
One of the things I was struggling with at first is seeing how all the pieces fit together. This might help:

```suggestion
    # Collect any entry points from the setup_py() object. Note that this was 
    # already normalized in `python_artifact.py`.
```

Eric-Arellano commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -589,15 +586,47 @@ async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
+        source_provides = setup_kwargs["entry_points"]
+
+    # Gather entry points with source description for any error messages when merging them.
+    entry_point_sources = {
+        f"{exported_addr} `provides.with_binaries()`": {"console_scripts": source_with_binaries},
+        f"{exported_addr} `entry_points`": source_entry_points,
+        f"{exported_addr} `provides.entry_points`": source_provides,
+    }
```
Nice!

kaos replied:
I'm trying to work out in my head in which order to have these, though. As I guess it will affect the potential error message in case you get merge conflicts..

Eric-Arellano commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -978,6 +1007,40 @@ def has_args(call_node: ast.Call, required_arg_ids: Tuple[str, ...]) -> bool:
+    *all_entry_points_with_descriptions_of_source: Tuple[str, Dict[str, Dict[str, str]]]
+) -> Dict[str, List[str]]:
+    """Merge all entry points, throwing ValueError if there are any conflicts."""
+    # keep source_name info for each entry_point until we have merged all
+    # sources, so we can provide better error messages in case of conflicts
```
I think you can probably get rid of this comment. It's not clear what `source_name` refers to, and I didn't anticipate you're going to remove the `description_of_source` - presumably it will be used, which this comment is saying fwict.

kaos replied:
yea, source_name could perhaps better be described as "description_of_source" to reflect the naming of the input arg. Was working on rephrasing the comment, but gave up ;) 
If the code is easy enough to follow, it's not needed any way..

Eric-Arellano commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -978,6 +1007,40 @@ def has_args(call_node: ast.Call, required_arg_ids: Tuple[str, ...]) -> bool:
+    # keep source_name info for each entry_point until we have merged all
+    # sources, so we can provide better error messages in case of conflicts
+    merged = cast(
+        # this gives us a two level deep defaultdict with the inner values being of list type
+        Dict[str, Dict[str, List[Tuple[str, str]]]],
```
Ah, this type hint helps a ton! Imo you can get rid of the comment, the type hint is self-documenting.

Eric-Arellano replied:
Type hint wise, I think you want to use `typing.DefaultDict`.

kaos replied:
> Ah, this type hint helps a ton! Imo you can get rid of the comment, the type hint is self-documenting.
you sure? the type hint was there all along :D

Eric-Arellano commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -978,6 +1007,40 @@ def has_args(call_node: ast.Call, required_arg_ids: Tuple[str, ...]) -> bool:
+        defaultdict(partial(defaultdict, list)),
+    )
+
+    for source_name, source_entry_points in all_entry_points_with_descriptions_of_source:
```
```suggestion
    for description_of_source, entry_points in all_entry_points_with_descriptions_of_source:
```

Eric-Arellano commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -978,6 +1007,40 @@ def has_args(call_node: ast.Call, required_arg_ids: Tuple[str, ...]) -> bool:
     return False
 
 
+def merge_entry_points(
```
Great, this function is now clear to me :) thanks for making those changes!

Eric-Arellano commented on src/python/pants/backend/python/macros/python_artifact_test.py:
```diff
@@ -0,0 +1,73 @@
+        ),
+    ],
+)
+def test_normalize_entry_points(entry_points, normalized, expect):
```
Thanks!!

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +153,160 @@ async def inject_pex_binary_entry_point_dependency(
+) -> Generator[Tuple[bool, str, str, str], None, None]:
+    """Looks at each entry point to see if it is a target address or not.
+
+    Returns a tuple of tuples: IsTarget, section, name, ref.
```
```suggestion
    Yields a tuple: is_target, section, name, ref.
```

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +153,160 @@ async def inject_pex_binary_entry_point_dependency(
+            yield ref.startswith(":") or "/" in ref, section, name, ref
+
+
+@rule(desc="Determining the entry points for a `python_distribution` target")
```
This isn't documented at all...you'll have to add `level=LogLevel.DEBUG` to this for it to show up in the dynamic UI. Otherwise the description gets kind of discarded.

Imo, this is a great description you wrote and helpful to be in the dynamic UI :)

kaos replied:
copy paste from `resolve_pex_entry_point`. I'll add the level there too, then..

Eric-Arellano commented on src/python/pants/backend/python/target_types_test.py:
```diff
@@ -364,35 +366,90 @@ def test_parse_requirements_file() -> None:
+        ),
+    )
+
+    def assert_injected(address: Address, *expected: Address) -> None:
```
Imo, more clear is to use `list[Address]` for expected. I think it better expresses that the first arg is different than the rest.

kaos replied:
ok. I was contemplating that...

Eric-Arellano commented on src/python/pants/backend/python/target_types_test.py:
```diff
@@ -364,35 +366,90 @@ def test_parse_requirements_file() -> None:
+        dedent(
+            """\
+            python_requirement_library(
+                name='ansicolors',
+                requirements=['ansicolors'],
+                module_mapping={'ansicolors': ['colors']},
```
What I mean with testing colors is to add an entry point like this: `colors_ep = "colors"`, such that your dep inference rule should pick up a dep on the `:ansicolors` target.

kaos replied:
Done. https://github.com/pantsbuild/pants/pull/11872/files#diff-a2ccc3f505ec7e33785464002b3c6192b325d97fcdcda65ad11876f2b5d7a500R423-R431 :)

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +153,160 @@ async def inject_pex_binary_entry_point_dependency(
+                address,
+                import_reference="module",
+                context=(
+                    f"The distribution target {address} has the field "
+                    f"`entry_points={field_str}`"
```
Even better is something like:

https://github.com/pantsbuild/pants/blob/7362f6a336845c1c987324def19aee3179364a00/src/python/pants/backend/python/target_types_rules.py#L126-L130

kaos replied:
They result in the same, except I didn't have the last part to which py module it mapped. Added.

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +153,160 @@ async def inject_pex_binary_entry_point_dependency(
+                owners.unambiguous
+                or ((maybe_disambiguated,) if maybe_disambiguated else Addresses())
+            )
+            entry_point_addresses += unambiguous_owners  # type: ignore[assignment]
```
I don't think this will work. `Addresses` is immutable. Instead, you need to reassign it.

kaos replied:
hehe, maybe it is my C background shining through. But this is reassignment. `a += b` translates to `a = a + b`. ;)

kaos replied:
But lets ensure there is a test, that proves it...

kaos replied:
Also, MyPy agrees...
```
src/python/pants/backend/python/target_types_rules.py:307:13: error:
Incompatible types in assignment (expression has type "Tuple[Address, ...]",
variable has type "Addresses")  [assignment]
```

Eric-Arellano replied:
Hm, to avoid confusion, how about using `module_owners = OrderedSet()` at the start of this `for` loop, then calling `.update()`(?) on it each iteration. After the loop, set `entry_point_addresses` equal to `entry_points.pex_binary_addresses` + `module_owners`.

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +153,160 @@ async def inject_pex_binary_entry_point_dependency(
+    for target in targets:
+        if not target.has_field(PexEntryPointField):
+            raise InvalidFieldException(
+                f'The "entry_points" target address {target.address} does not resolve '
+                f"to a pex_binary, but: {target.alias}"
```
> All target addresses in the `entry_points` field must be for `pex_binary` targets, but the target `{address}` includes the value `{target.address}`, which has the target type `{target.alias}`.
>
> Alternatively, you can use a module like `project.app:main`. See {docs_url('python-distributions'}).

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +153,160 @@ async def inject_pex_binary_entry_point_dependency(
+
+
+@rule(desc="Determining the entry points for a `python_distribution` target")
+async def resolve_python_distribution_entry_points(
```
Nice, good job parallelizing this rule :)

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +153,160 @@ async def inject_pex_binary_entry_point_dependency(
+
+    entry_points: dict[str, dict[str, EntryPoint]] = defaultdict(dict)
+
+    # parse refs/replace with resolved pex entry point, and validate console entry points have function
```
Style nit: comments should be sentences.

```suggestion
    # Parse refs/replace with resolved pex entry point, and validate console entry points have function.
```

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +153,160 @@ async def inject_pex_binary_entry_point_dependency(
+    for is_target, section, name, ref in classified_entry_points:
+        if is_target:
+            entry_point = binary_entry_point_by_address[address_by_ref[ref]].val
+            # ResolvedPexEntryPoint.val is Optional[EntryPoint]
+            if entry_point is None:
+                continue
```
Interesting, I'm honestly not sure what the right behavior is here. I think probably to error? It doesn't make sense to depend on a `pex_binary` with `entry_point='<none>'`, as that means there is no entry point. That seems more like an unintentional mistake.

kaos replied:
So, perhaps warn? In case it indeed was intentional (even just for testing purposes..)

Eric-Arellano reviewed (commented):
Nice, thanks so much for iterating on this Andreas! This is shaping up really nicely.

kaos commented on src/python/pants/backend/python/target_types.py:
```diff
@@ -724,6 +725,54 @@ def compute_value(cls, raw_value: Optional[PythonArtifact], address: Address) ->
+class PythonDistributionEntryPoints(NestedDictStringToStringField, AsyncFieldMixin):
+    alias = "entry_points"
+    required = False
+    help = dedent(
```
I'm particularly grateful for your nit-picks. I'm used to being the nitty one :p so it's refreshing to be on the receiving end for a change :D 

I'll blame it on that I'm spending all my brain cells on the actual implementation, so there's not much left to make it pretty (true or not!). So happy you can step in there ;) (so, I expect it'll get better over time.. as I get more familiar with the code base and style)

kaos commented on src/python/pants/backend/python/goals/setup_py.py:
```diff
@@ -589,15 +586,47 @@ async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
+            ),
+        )
+
+        if entry_points_field.val:
```
good catch.

Eric-Arellano commented on src/python/pants/backend/python/target_types.py:
```diff
@@ -724,6 +725,93 @@ def compute_value(cls, raw_value: Optional[PythonArtifact], address: Address) ->
+# See `target_type_rules.py` for the `Resolve..Request -> Resolved..` rule
+@dataclass(frozen=True)
+class ResolvedPythonDistributionEntryPoints:
+    val: FrozenDict[str, FrozenDict[str, PythonDistributionPexableEntryPoint]] = FrozenDict()
```
I forgot what this was supposed to be. This might make it more clear:

```suggestion
    categories_to_entry_points: FrozenDict[str, FrozenDict[str, PythonDistributionPexableEntryPoint]] = FrozenDict()
```

Maybe add a comment: "E.g. `{"console_scripts": {"ep": PythonDistributionEntryPoint(...)}}`.

Eric-Arellano commented on src/python/pants/backend/python/target_types.py:
```diff
@@ -724,6 +725,93 @@ def compute_value(cls, raw_value: Optional[PythonArtifact], address: Address) ->
+
+
+@dataclass(frozen=True)
+class PythonDistributionPexableEntryPoint:
+    """Pexable means that the entry point could be from a pex_binary target."""
```
There's no other type of `PythonDistribution`, so imo you can remove `Pexable`.
```suggestion
class PythonDistributionEntryPoint:
    """Note that this stores if the entry point comes from an address to a `pex_binary` target."""
```

Eric-Arellano replied:
If you make this rename, you can rename `PythonDistributionEntryPoints` (plural) to `PythonDistributionEntryPointsField` maybe.

kaos replied:
Yeah, I avoided PythonDistributionEntryPoint due to there being a plural form, so a mere `s` to distinguish them by is too little imo. Also, I felt that the name `PythonDistributionEntryPointsField` got unwieldy long.. and also didn't quite follow the naming pattern for the other fields.. that doesn't have the `Field` suffix. Maybe we should start using `Field` suffix for all fields, to be consistent.. ? (it helps making it clearer when the type shows up else where, that it is a field it refers to.. which isn't always obvious, in my eyes at least)

Eric-Arellano replied:
> Maybe we should start using Field suffix for all fields, to be consistent.. ?

Probably. We do use `Field` in a couple names if you ripgrep it, but you're right that not many yet.

kaos replied:
But if you prefer that, I'm happy to change it...

Eric-Arellano commented on src/python/pants/backend/python/target_types.py:
```diff
@@ -724,6 +725,93 @@ def compute_value(cls, raw_value: Optional[PythonArtifact], address: Address) ->
+                        if not pexable.pex_binary_address
+                    }
+                )
+                for key, value in self.val.items()
```
Key and value are vague, it's not clear what these actually mean. Maybe `category` and `entry_points`? Then, `ep_name` and `ep_val` for the inner loop?

Eric-Arellano commented on src/python/pants/backend/python/target_types.py:
```diff
@@ -724,6 +725,93 @@ def compute_value(cls, raw_value: Optional[PythonArtifact], address: Address) ->
+        """Returns the addresses to all pex binary targets owning entry points used."""
+        return Addresses(
+            pexable.pex_binary_address
+            for key, value in self.val.items()
+            for _, pexable in value.items()
```
Ditto on renaming

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +159,174 @@ async def inject_pex_binary_entry_point_dependency(
+
+    # Pick out all target addresses up front, so we can use MultiGet later.
+    #
+    # This calls for a bit of trickery however (using the "y_by_x" mapping dicts), so we keep track
+    # of which address belongs to which entry point.
```
Stale?

kaos replied:
No, it still applies. `y_by_x` refers to the `address_by_ref` and `binary_entry_point_by_address` respectively. Maybe I should spell them out instead.

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +159,174 @@ async def inject_pex_binary_entry_point_dependency(
+    # This calls for a bit of trickery however (using the "y_by_x" mapping dicts), so we keep track
+    # of which address belongs to which entry point.
+    target_refs = [ref for is_target, _, _, ref in classified_entry_points if is_target]
+    # intermediate step, as Get(Targets) returns a deduplicated set..
+    target_addresses = await Get(
+        Addresses, UnparsedAddressInputs(target_refs, owning_address=address)
+    )
+    address_by_ref = dict(zip(target_refs, target_addresses))
+    targets = await Get(Targets, Addresses, target_addresses)
```
I'm not following this part. Why not go right away to `await Get(Targets, UnparsedAddressInputs)`?

kaos replied:
Did you miss my comment, or maybe it didn't explain it well enough?
If I go right away to `Get(Targets)`, I don't have a 1:1 mapping from my inputs, in case of duplicates in the input, so can not know which unparsed address represents which target (which was the case in the test data, luckily!)

kaos replied:
I've added a few more words to that comment...

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +159,174 @@ async def inject_pex_binary_entry_point_dependency(
+        entry_points[section][name] = PythonDistributionPexableEntryPoint(entry_point, owner)
+
+    return ResolvedPythonDistributionEntryPoints(
+        FrozenDict({key: FrozenDict(value) for key, value in entry_points.items()})
```
Ditto renaming to `category` and `entry_points`.

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +159,174 @@ async def inject_pex_binary_entry_point_dependency(
+            entry_point = binary_entry_point_by_address[owner].val
+            if entry_point is None:
+                logger.warning(
+                    f"The entry point {name} in {section} references a pex binary {ref}, "
```
```suggestion
                    f"The entry point {name} in {section} for {address} references a pex binary {ref} "
```

Eric-Arellano replied:
And then add a "So what?" like

> Skipping this entry because '<none' is not valid as an entry point.

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +159,174 @@ async def inject_pex_binary_entry_point_dependency(
+        address = original_tgt.target.address
+        all_module_entry_points = [
+            (key, name, entry_point)
+            for key, section in entry_points.explicit_modules.items()
```
Ditto `category` and `entry_points`.

Eric-Arellano commented on src/python/pants/backend/python/target_types_rules.py:
```diff
@@ -143,29 +159,174 @@ async def inject_pex_binary_entry_point_dependency(
     )
-    return InjectedDependencies(addresses)
+
+    if not entry_points.val:
+        entry_point_addresses = Addresses()
+    else:
```
I think you can skip this `if` block for less indentation. If there are no vals, the below code will simply have nothing to do.

kaos replied:
True, remnant from when `val` was possibly `None`..

Eric-Arellano reviewed (approved):
Yay! The functionality and tests look great, now only a couple style things

Eric-Arellano commented on src/python/pants/backend/python/target_types_test.py:
```diff
@@ -364,35 +366,109 @@ def test_parse_requirements_file() -> None:
+                    }
+                },
+            )
+
```
(In a followup) Could you please add a test for dep inference of entry points defined in the `provides=setup_py()`?

kaos replied:
Sure thing. But trying to remember what's changed since: https://github.com/pantsbuild/pants/pull/11872#discussion_r613511687
> [...] you can still use the old setup_py() way of doing it, only you'll miss out a) dependency inference

Hmm... looking at the `inject_python_distribution_dependencies` I don't think that we infer dependencies on entry_points from `setup_py()` at all.. so that test should not work ;)

kaos replied:
But then again, that was the point of #11807 which supposedly was incorporated into this PR... eeh.. a little too much water under the bridge from start to finnish, I'd say..

kaos replied:
So, question becomes, do we want to spend time fixing that dep inference, or simply leave be waiting for #12410

Eric-Arellano replied:
Ah, okay. Makes sense.

Totally depends on what you have the time and interest to do: #12410 is more of a blocker b/c right now we have two Pants-provided ways of doing something, and we want to get rid of the `.with_binaries()` tech debt.

It would also be great to add dep inference for entry points defined inside a `setup_py()`, but that isn't a blocker. It's a new feature. If you have the time and interest, that'd be great - okay if not. Either way, maybe create a issue to track this so we don't forget it's a feature request?

kaos replied:
I'll post a PR deprecating `.with_binaries()` tomorrow.
Can also look into the dep inference on setup_py entry_points, shouldn't be too difficult, now :)

kaos replied:
Deprecation PR: #12413

kaos replied:
Dep inference for `setup_py()` entry points PR: #12414

Eric-Arellano reviewed (approved):
Excellent contribution, @kaos 🥳

This is a tricky part of the codebase, and you dove right in :) I'm really excited with how this turned out.

One small followup if you're willing, but I'm going to merge this because that followup doesn't block and this is an excellent change as is.
